adam/perception/perception_graph_predicates.py

```diff
@@ -0,0 +1,644 @@
+        return graph_node[0]
+    else:
+        return graph_node
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class AxisPredicate(NodePredicate):
+    """
+    Represents constraints on an axis given in a `PerceptionGraphPattern`
+    """
+
+    curved: Optional[bool] = attrib(validator=optional(instance_of(bool)))
+    directed: Optional[bool] = attrib(validator=optional(instance_of(bool)))
+    aligned_to_gravitational: Optional[bool] = attrib(
+        validator=optional(instance_of(bool))
+    )
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        unwrapped_graph_node = unwrap_if_necessary(graph_node)
+
+        if isinstance(unwrapped_graph_node, GeonAxis):
+            if self.curved is not None and self.curved != unwrapped_graph_node.curved:
+                return False
+            if (
+                self.directed is not None
+                and self.directed != unwrapped_graph_node.directed
+            ):
+                return False
+            if (
+                self.aligned_to_gravitational is not None
+                and self.aligned_to_gravitational
+                != unwrapped_graph_node.aligned_to_gravitational
+            ):
+                return False
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def from_axis(axis_to_match: GeonAxis) -> "AxisPredicate":
+        return AxisPredicate(
+            curved=axis_to_match.curved,
+            directed=axis_to_match.directed,
+            aligned_to_gravitational=axis_to_match.aligned_to_gravitational,
+        )
+
+    def dot_label(self) -> str:
+        constraints = []
+
+        if self.curved is not None:
+            constraints.append(f"{sign(self.curved)}curved")
+        if self.directed is not None:
+            constraints.append(f"{sign(self.directed)}directed")
+        if self.aligned_to_gravitational is not None:
+            constraints.append(f"{sign(self.aligned_to_gravitational)}grav_aligned")
+
+        return f"axis({', '.join(constraints)})"
+
+    def is_equivalent(self, other) -> bool:
+        if isinstance(other, AxisPredicate):
+            return (
+                self.aligned_to_gravitational == other.aligned_to_gravitational
+                and self.curved == other.curved
+                and self.directed == other.directed
+            )
+        return False
+
+    def matches_predicate(self, predicate_node: "NodePredicate") -> bool:
+        if isinstance(predicate_node, AxisPredicate):
+            if self.curved != predicate_node.curved:
+                return False
+            if self.directed != predicate_node.directed:
+                return False
+            if self.aligned_to_gravitational != predicate_node.aligned_to_gravitational:
+                return False
+            else:
+                return True
+        else:
+            return False
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class GeonPredicate(NodePredicate):
```

Comment:
Most of the code like this class is moved from perception_graph.py, right?

Reply:
Correct, I just relocated these predicates into the new file.